### ISSUE TITLE
Excel error message

### ISSUE THREAD

**Issue by GuyVervilleSpiria** [shuchkin/simplexlsxgen#24]
Hello, 
First of all thank you for the excellent class! Very easy to use. I have a curious problem. When I test my code locally, I can read the Excel file correctly. However, testing from my staging environment, I get an error message from Excel when I try to read the downloaded file. The data is all there, but Excel tells me that the file needs to be repaired.
[SN482941614375363.xlsx](https://github.com/shuchkin/simplexlsxgen/files/6052633/SN482941614375363.xlsx)

Here is an example file.

**Comment by shuchkin** [comment]
try debug
https://github.com/shuchkin/simplexlsxgen#debug

**Comment by GuyVervilleSpiria** [comment]
This simply gives me a file with a two columns and I get the same error from Excel. _Numbers_ has no problem opening the file... Very strange.

[debug.xlsx](https://github.com/shuchkin/simplexlsxgen/files/6055085/debug.xlsx)

**Comment by shuchkin** [comment]
show print_r( $data );
I want check your input data structure

**Comment by GuyVervilleSpiria** [comment]
Nothing fancy here. Here is var_dump. The first line is the header.
`array(2) {
  [0]=>
  array(14) {
    [0]=>
    string(4) "Ref."
    [1]=>
    string(14) "User Reference"
    [2]=>
    string(4) "Name"
    [3]=>
    string(2) "ID"
    [4]=>
    string(15) "Date expiration"
    [5]=>
    string(13) "Date creation"
    [6]=>
    string(6) "Statut"
    ["fever-symptoms"]=>
    string(19) "Sympômes fiévreux"
    ["do-you-have-a-cough"]=>
    string(4) "Toux"
    ["trouble-breathing"]=>
    string(19) "Trouble à respirer"
    ["loss-of-smell"]=>
    string(19) "Perte de l’odorat"
    ["general-symptoms"]=>
    string(21) "Symptômes généraux"
    ["nausea"]=>
    string(8) "Nausées"
    ["sore-throat"]=>
    string(13) "Maux de gorge"
  }
  [1]=>
  array(14) {
    [0]=>
    string(6) "211601"
    [1]=>
    string(5) "53243"
    [2]=>
    string(21) "employe_un employe_un"
    [3]=>
    string(24) "employe_un_webspiria.com"
    [4]=>
    string(16) "26/02/2021 11:04"
    [5]=>
    string(16) "27/02/2021 07:04"
    [6]=>
    string(3) "✕"
    ["fever-symptoms"]=>
    string(3) "Oui"
    ["do-you-have-a-cough"]=>
    string(3) "Non"
    ["trouble-breathing"]=>
    string(3) "Non"
    ["loss-of-smell"]=>
    string(3) "Oui"
    ["general-symptoms"]=>
    string(3) "Oui"
    ["nausea"]=>
    string(3) "Non"
    ["sore-throat"]=>
    string(3) "Non"
  }
}`

I will try the PhpSpreadsheet script to see if I have the same problem. The Excel file was test by three other persons, and they got all the Repair request by Excel.

**Comment by GuyVervilleSpiria** [comment]
My code is as follows :
`
foreach ($versionData as $key => $versionDatum) {
            $data[] = [...] / code for generating the header/;
// a loop because there can be numerous sheets.
            foreach ($versionDatum as $versionItem) {
                if (empty($versionItem["results"])) {
                    $results = $this->fillWithZeros(count($headers[$key][0]));
                } else {
                    $results = $versionItem["results"];
                }
                $data[] = array_merge([
                    $versionItem['id'],
                    $versionItem['user_id'],
                    $companyUsers[$versionItem['user_id']]["fullname"],
                    $companyUsers[$versionItem['user_id']]["name"],
                    date_format(date_create($versionItem['created_at']), $dateFormat),
                    date_format(date_create($versionItem['expire_at']), $dateFormat),
                    $versionItem['clearance']
                ], $results);
            }
            $xlsx->addSheet($data, $headers[$key][1]);
            $data = [];
        }
        return $xlsx->downloadAs($companyId . time() . '.xlsx');
`

**Comment by GuyVervilleSpiria** [comment]
For your information, I tried the library PHPSpreadsheet with no problem with the same data.

**Comment by shuchkin** [comment]
GuyVervilleSpiria your bad, plz use exit() after downloadAs(...) to stop output 😡 
<img width="1091" alt="Screenshot_440" src="https://user-images.githubusercontent.com/315872/109427656-95c09a80-7a1d-11eb-8697-8937c10816b0.png">

**Comment by GuyVervilleSpiria** [comment]
This is a function which return the output. The exit() comes after that. It is there.

**Comment by shuchkin** [comment]
don't "echo" this function

**Comment by GuyVervilleSpiria** [comment]
Ok thanks

Guy Verville
SPIRIA
________________________________
De : Sergey Shuchkin <notifications@github.com>
Envoyé : Sunday, February 28, 2021 12:59:58 PM
À : shuchkin/simplexlsxgen <simplexlsxgen@noreply.github.com>
Cc : Guy Verville <gverville@spiria.com>; Author <author@noreply.github.com>
Objet : Re: [shuchkin/simplexlsxgen] Excel error message (#24)


don't "echo" this function

**Comment by shuchkin** [comment]
"This is a function which return the output." --- No, downloadAs() is output method

if u want $xlsx content (ZIP) as string, just write $xlsx_content = (string) $xlsx;

**Comment by fernando2amigos** [comment]
Hi @shuchkin.
I'm having random issues with SOME generated files. 
Particularly this one is receiving complaints from Microsoft Office, however I can open it without errors in WPS Suite and also Google Sheets.
![image](https://user-images.githubusercontent.com/69151149/127048572-2231b5de-2b0a-48bd-b1c7-87c5e4babe63.png)
I've checked above issue having content after your _thank you_ message but there's nothing after that. This doesn't happens to ALL files, but very few. Unfortunately haven't found a pattern yet. 

Where else could I look for issues? What's the way to debug XML structure validity?

PS: Thanks fro your great effort and job with the library!

**Comment by shuchkin** [comment]
Fernando, send me your xlsx to sergey.shuchkin@gmail.com

**Comment by fernando2amigos** [comment]
@shuchkin I've sent you the file yesterday. I appreciate your willingness.

I've already solved the issue with sheet names making the file unable to be opened. It was slashes and or ampersands.
So I've made this function to return safe sheet names.

```php
    /**
     * Strip out invalid chars and encode special ones into XML-valid entities
     *
     * @param string $name
     * @return string
     */
    public static function xlsxSafeSheetName(string $name): string
    {
        // Names longer than 31 chars get cut when shown
        if (strlen($name) > 31) {
            $name = substr($name, 0, 28) . '...';
        }

        // Strip out invalid Microsoft Excel chars
        $name = str_replace(
            [':','\\','/','?','*','[',']'],
            [' ',' ' ,'-',' ',' ',' ',' '],
            $name
        );

        return htmlspecialchars($name, ENT_XML1);
    }
```

However, some sheets still have problems and their names turn into "Recovered_sheetX". 
These files, being reports only contains strings and integers. No formulas nor special stuff at all.
I'll try to get one without sensible information and send it to you again.

**Comment by shuchkin** [comment]
try 1.0.17

**Comment by fernando2amigos** [comment]
Just tried with 1.0.16 and tried to open in Office 2007 which gave me a good insight: The repair message mentioned something about removing hyperlinking feature on worksheet 27.
After some inspection I believe to have found the problem.

It's creating `<hyperlink>` tags with empty `location` value:
```xml
<hyperlinks>
  <hyperlink ref="B2" r:id="rId1" location="" display="mailto:xxxx@xxxx"/> 
  <hyperlink ref="B3" r:id="rId2" location="" display="mailto:yyyyy@yyyy"/>
  <hyperlink ref="E3" r:id="rId3" location="" display="https://www.facebook.com/blablabla/"/>
  <hyperlink ref="B4" r:id="rId4" location="" display="mailto:zzzzz@zzzzzz"/>
...
</hyperlinks>
```
However, 
```xml
<c r="B2" t="s" s="1">
  <v>299</v>
</c>
...
<c r="B3" t="s" s="1">
  <v>301</v>
</c>
...
<c r="B4" t="s" s="1">
<v>304</v>
</c>
```
Further, I'm not creating hyperlinks from data. They're just plain strings.
I'll send you this sheet over email for your inspection.

**Comment by m-genser** [comment]
I am also experiencing this excel error. 
After some debugging, I realized thats triggered when I add 'Ö', 'Ä', or 'Ü' to the spreadsheet data payload.
When I use mb_convert_encoding() to change this chars to question marks it works without any problem.

I am currently playing around with throwing different encodings at the library and I'll update here when I found a solution. (I am sure that someone with the same problem will end up here :) )
